tests/src/Kernel/UniqueEmailTest.php

```diff
@@ -0,0 +1,586 @@
+  }
+
+  /**
+   * Test the unique email validator in 'include' mode with an existing decoupled
+   * user with no matching roles.
+   *
+   * @covers ::validate
+   */
+  public function testModeIncludeDecoupledWithoutRole() {
+    $role = $this->createRole([]);
+    $this->setUniqueEmailsConfig('include', [$role]);
+
+    $existing_user = $this->createDecoupledUser();
+
+    // Test validating a decoupled user without the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user without role passes validation.');
+
+    // Test validating a decoupled user with the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $user->addRole($role);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user with role passes validation.');
+
+    // Test validating a coupled user without the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(FALSE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Coupled user without the role passes validation.');
+
+    // Test validating a coupled user with the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(FALSE, $existing_user->email_prefix);
+    $user->addRole($role);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Coupled user with the role passes validation.');
+  }
+
+  /**
+   * Test the unique email validator in 'include' mode with an existing decoupled
+   * user with a matching roles.
+   *
+   * @covers ::validate
+   */
+  public function testModeIncludeDecoupledWithRole() {
+    $role = $this->createRole([]);
+    $this->setUniqueEmailsConfig('include', [$role]);
+
+    $existing_user = $this->createDecoupledUser();
+    $existing_user->addRole($role);
+    $existing_user->save();
+
+    // Test validating a decoupled user without the role.
+    // Expected result: No violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertEmpty($violations, 'Decoupled user without role passes validation.');
+
+    // Test validating a decoupled user with the role.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(TRUE, $existing_user->email_prefix);
+    $user->addRole($role);
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Decoupled user with role fails validation.');
+
+    // Test validating a coupled user without the role.
+    // Expected result: Violations
+    $user = $this->createUnsavedUser(FALSE, $existing_user->email_prefix);
+    $violations = $user->validate();
+    $this->assertNotEmpty($violations, 'Coupled user without the role fails validation.');
```

Comment:
As the coupled user has the role, and the decoupled user doesn't, shouldn't thise be assertEmpty.